core/preconditioner/isai.cpp

```diff
@@ -0,0 +1,146 @@
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/preconditioner/isai.hpp>
+
+
+#include <functional>
+#include <memory>
+#include <type_traits>
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/utils.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+
+
+#include "core/preconditioner/isai_kernels.hpp"
+
+
+namespace gko {
+namespace preconditioner {
+namespace isai {
+
+
+GKO_REGISTER_OPERATION(generate_l_inverse, isai::generate_l_inverse);
+GKO_REGISTER_OPERATION(generate_u_inverse, isai::generate_u_inverse);
+
+
+}  // namespace isai
+
+
+/**
+ * Helper function that converts the given matrix to the (const) CSR format with
+ * additional sorting.
+ *
+ * If the given matrix was already sorted, is on the same executor and with a
+ * dynamic type of `const Csr`, the same pointer is returned with an empty
+ * deleter.
+ * In all other cases, a new matrix is created, which stores the converted Csr
+ * matrix.
+ * If `skip_sorting` is false, the matrix will be sorted by column index,
+ * otherwise, it will not be sorted.
+ */
+template <typename Csr>
+std::unique_ptr<const Csr, std::function<void(const Csr *)>>
+convert_to_csr_and_sort(std::shared_ptr<const Executor> &exec, const LinOp *mtx,
+                        bool skip_sorting)
+{
+    static_assert(
+        std::is_same<Csr, matrix::Csr<typename Csr::value_type,
+                                      typename Csr::index_type>>::value,
+        "The given `Csr` type must be of type `matrix::Csr`!");
+    if (skip_sorting && exec == mtx->get_executor()) {
+        auto csr_mtx = dynamic_cast<const Csr *>(mtx);
+        if (csr_mtx) {
+            // Here, we can just forward the pointer with an empty deleter
+            // since it is already sorted and in the correct format
+            return {csr_mtx, null_deleter<const Csr>{}};
+        }
+    }
+    auto copy = Csr::create(exec);
+    as<ConvertibleTo<Csr>>(mtx)->convert_to(copy.get());
+    // Here, we assume that a sorted matrix converted to CSR will also be
+    // sorted
+    if (!skip_sorting) {
+        copy->sort_by_column_index();
+    }
+    return {copy.release(), std::default_delete<const Csr>{}};
+}
+
+
+template <isai_type IsaiType, typename ValueType, typename IndexType>
```

Comment:
Does it make sense to instantiate this when `IsaiType = isai_type::upper` ?

Reply:
It should not hurt since it is a non-`virtual` function. Additionally, this would complicate the code quite a lot more (I can only come up with a method that would probably be similar to writing both `LowerIsai` and `UpperIsai` by hand), and since it is a `private` function, I did not see the need for that.